### PR TITLE
Jk/log grad norms/log grad norms

### LINKED ISSUE #688: Log gradient norms to understand and debug model better
### Is your feature request related to a problem? Please describe.

See title, ideally with some plotting code, and potentially behind a flag to prevent large metric files.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by Jubeku**
## Description

This PR is based on @sophie-xhonneux's [log_grad_norm](https://github.com/ecmwf/WeatherGenerator/tree/sophiex/dev/log-grad-norms) branch in #685, modified to allow logging gradients when running in parallel on multiple GPUs with FSDP2. 

## Issue Number

Closes #688 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -914,6 +923,31 @@ def _log(self, stage: Stage):
...
+    def _log_instant_grad_norms(self, stage: Stage, total_norm):
+        """
+        Log instantaneous grad norms, we do not average because of the cost and because we want to
+        measure the actual values
+
+        TODO test DDP case
+        """
+        self.last_grad_norm = (
+            total_norm.full_tensor().item() if self.cf.world_size > 1 else total_norm.item()
+        )
```
Comment (Jubeku) [inline src/weathergen/train/trainer.py]: As mentioned [here](https://www.linkedin.com/posts/kai-zhang-53910214a_dtensor-with-normpartial-stores-local-partial-activity-7360490130744266753-2EhX), `full_tensor().item()` is needed in parallel runs with FSDP2. 

I tested this by logging both ways of calculating:
```
000 : 00010/02048 : 000010 : loss = 1.0287E+00 (lr=1.64E-06, gradient norm=0.983, gradient norm FT=1.403, s/sec=0.236)

ERA5 : 1.0287E+00 


000 : 00020/02048 : 000020 : loss = 1.0101E+00 (lr=3.34E-06, gradient norm=0.587, gradient norm FT=0.817, s/sec=0.435)

ERA5 : 1.0101E+00 
```

File: src/weathergen/train/trainer.py [hunk 2]
```diff
@@ -914,6 +923,31 @@ def _log(self, stage: Stage):
...
+        grad_norms = {"total_grad_norm": self.last_grad_norm}
+        for name, param in self.model.named_parameters():
+            if param.grad is not None:
+                # grad_norms["grad_norm_" + name] = param.grad.norm().item()
+                grad_norms["grad_norm_" + name] = (
+                    param.grad.norm().full_tensor().item()
+                    if self.cf.world_size > 1
+                    else param.grad.norm().item()
+                )
+
```
Comment (Jubeku) [inline src/weathergen/train/trainer.py:1020]: Same as above, we also need `.full_tensor().item()` here in multi-gpu mode. 

Tested it by printing both versions on 2 GPUs:
```
print(".item():", param.grad.norm().item())
print(".full_tensor().item()", param.grad.norm().full_tensor().item())

.item(): 0.028306283056735992
.item(): 0.022433193400502205
.full_tensor().item() 0.03611777722835541
.full_tensor().item() 0.03611777722835541
```

File: src/weathergen/train/trainer.py [hunk 3]
```diff
@@ -914,6 +923,31 @@ def _log(self, stage: Stage):
...
+            total_norm.full_tensor().item() if self.cf.world_size > 1 else total_norm.item()
+        )
+        grad_norms = {"total_grad_norm": self.last_grad_norm}
+        for name, param in self.model.named_parameters():
+            if param.grad is not None:
+                # grad_norms["grad_norm_" + name] = param.grad.norm().item()
+                grad_norms["grad_norm_" + name] = (
+                    param.grad.norm().full_tensor().item()
+                    if self.cf.world_size > 1
+                    else param.grad.norm().item()
```
Comment (tjhunter) [inline src/weathergen/train/trainer.py]: shouldn't you divide by the number of items in the gradient? otherwise, if every component in the gradient is equal, you are biased by batching computations

File: config/default_config.yml [hunk 4]
```diff
@@ -10,7 +10,7 @@ embed_dropout_rate: 0.1
 target_cell_local_prediction: True
 
 ae_local_dim_embed: 1024
-ae_local_num_blocks: 2
+ae_local_num_blocks: 0
```
Comment (sophie-xhonneux) [inline config/default_config.yml]: Can we as always revert back to the original default config please?

File: src/weathergen/train/trainer.py [hunk 5]
```diff
@@ -914,6 +923,31 @@ def _log(self, stage: Stage):
 
         self.loss_unweighted_hist, self.loss_model_hist, self.stdev_unweighted_hist = [], [], []
 
+    def _log_instant_grad_norms(self, stage: Stage, total_norm):
+        """
+        Log instantaneous grad norms, we do not average because of the cost and because we want to
+        measure the actual values
+
+        TODO test DDP case
```
Comment (sophie-xhonneux) [inline src/weathergen/train/trainer.py]: can you remove the TODO and copy your FSDP2 comment into the code. Thank you!

File: src/weathergen/train/trainer.py (hunk 3 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: Not sure I follow. But as far as I know the gradient norm logging is correct and people do not commonly account for the number of dimensions, as for batching this is handle in the forward pass and thus is automatically dealt with during backprop.

File: src/weathergen/train/trainer.py [hunk 6]
```diff
@@ -964,7 +999,7 @@ def _log_terminal(self, bidx: int, epoch: int, stage: Stage):
                     # samples per sec
                     dt = time.time() - self.t_start
                     pstr = "{:03d} : {:05d}/{:05d} : {:06d} : loss = {:.4E} "
-                    pstr += "(lr={:.2E}, s/sec={:.3f})"
+                    pstr += "(lr={:.2E}, gradient norm={:.3f}, s/sec={:.3f})"
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: We should have `gradient norm={:.3f}` only when gradient logging is enabled.

File: src/weathergen/train/trainer.py [hunk 7]
```diff
@@ -539,6 +541,7 @@ def train(self, epoch):
 
         # Unweighted loss, real weighted loss, std for losses that need it
         self.loss_unweighted_hist, self.loss_model_hist, self.stdev_unweighted_hist = [], [], []
+        self.last_grad_norm = 0.0
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: I think this won't be needed when `_log_terminal()` only prints when `log_grad_norm = True`. But let's please make sure it works in all variants.

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -942,6 +951,32 @@ def _log(self, stage: Stage):
...
+        """
+        self.last_grad_norm = (
+            total_norm.full_tensor().item() if self.cf.world_size > 1 else total_norm.item()
+        )
+        grad_norms = {"total_grad_norm": self.last_grad_norm}
+        for name, param in self.model.named_parameters():
+            if param.grad is not None:
+                # grad_norms["grad_norm_" + name] = param.grad.norm().item()
+                grad_norms["grad_norm_" + name] = (
+                    param.grad.norm().full_tensor().item()
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: We have this statement with `if self.cf.world_size > 1` at least twice, and probably more. Can we put it into a thin function `get_item()` or something.

**Comment by Jubeku** [comment]
All `grad_norm.`-metrics will be excluded from MLFlow upload through this PR: https://gitlab.jsc.fz-juelich.de/esde/WeatherGenerator-private/-/merge_requests/78 cc @tjhunter

File: src/weathergen/train/trainer.py [hunk 9]
```diff
@@ -89,6 +89,8 @@ def init(self, cf: Config, devices):
         # world_size gets overwritten by current setting during init_ddp()
         self.world_size_original = cf.get("world_size", None)
 
+        self.log_grad_norms = cf.get("log_grad_norms", False)
```
Comment (clessig) [inline src/weathergen/train/trainer.py:93]: Can we add this to default config but disabled.

**Review by clessig — APPROVED** [review]

File: src/weathergen/train/trainer.py (hunk 9 above)
Comment (Jubeku) [reply src/weathergen/train/trainer.py:93]: Done.
